Extract billing

blakepettersson opened the pull request:
Here are the changes required in order for aptible/dashboard.aptible.com#352 to work. This moves all Stripe-related attributes from `organization.js` to `billing-details.js`, and adds a new adapter and serializer for the billing details model.

mixonic commented on app/adapters/billing-detail.js:
```diff
@@ -0,0 +1,13 @@
+  host: config.billingBaseUri,
+
+  pathForType: function(type) {
+    const underscored = Ember.String.underscore(type);
```
`Ember` must be imported:

```
import Ember from "ember";
```

mixonic commented on app/models/billing-detail.js:
```diff
@@ -1,10 +1,19 @@
+  hasStripeCustomer: Ember.computed.bool('stripeCustomerId'),
+  hasStripe: Ember.computed.and('hasStripeCustomer', 'hasStripeSubscription'),
+  organization: DS.belongsTo('organization', {async:true})
+  //stripeSubscriptionStatus: DS.attr()
```
Can these comments simply be removed?

mixonic replied:
Also, we usually group model attributes and relationships before normal props

``` js
export default DS.Model.extend({
  plan: DS.attr('string'),
  foo: DS.attr('number'),
  organizations: DS.hasMany('organization', {async: true}),

  hasSomething: Ember.computed.bool('someThing')
});
```

It must make it slightly easier to reason about and understand the schema of the model.

mixonic commented on app/models/organization.js:
```diff
@@ -20,18 +20,18 @@ export default DS.Model.extend({
-  hasStripeSubscription: Ember.computed.bool('stripeSubscriptionId'),
-  hasStripeCustomer: Ember.computed.bool('stripeCustomerId'),
-  hasStripe: Ember.computed.and('hasStripeCustomer', 'hasStripeSubscription'),
+  billingDetail: DS.belongsTo('billingDetail', {async:true}),
```
All types passed to `belongsTo` and `hasMany` should be `dash-erized`, not `camelCase`.

mixonic commented on app/serializers/billing-detail.js:
```diff
@@ -0,0 +1,8 @@
+
+export default ApplicationSerializer.extend({
+  extract: function(store, type, payload, id, requestType) {
+    this.__requestType = requestType; // used by `extractMeta`
```
Can this comment be elaborated upon? Used by `extractMeta` for what? It seems unfortunate to set a `__doubleUnderscorePrivate` value.

skylar-anderson replied:
FYI @blakepettersson - Since we are overloading ember-data-hal-9000's serializer the requestType needs to be cached.  Explanation can be found here:  https://github.com/201-created/ember-data-hal-9000/blob/master/addon/serializer.js#L37

skylar-anderson replied:
@blakepettersson - actually, after reviewing dashboard and API, it looks like you are using `store.find('billingDetail', $id')` rather than trying to serialize a singleton resource.  Is this serializer necessary at all anymore?

blakepettersson replied:
@sandersonet currently it seems to work without the serializer, but I recall that I had to have it even when loading a single model since otherwise it broke horribly. There is also a test that currently fails when I remove the serializer, I'm not sure why.

mixonic commented on tests/unit/models/billing-detail-test.js:
```diff
@@ -0,0 +1,33 @@
+
+test('it exists', function() {
+  var model = this.subject();
+  // var store = this.store();
```
comment, can be removed

mixonic commented:
Looks like there are a lot of stray comments that can be removed now, and the branch and probably be rebased to lose those `WIP` commits as well.

Heading in a good direction though! You got the tricky bits about the adapter dead right.

skylar-anderson commented on app/models/organization.js:
```diff
@@ -20,18 +20,18 @@ export default DS.Model.extend({
   plan: DS.attr('string'),
```
Can this be removed?

blakepettersson commented:
OK, so I rebased the branch. I'm not sure if the serializer is actually needed; I'll take a look at the failing test in dashboard to see why its failing.

blakepettersson committed:
Move Stripe attributes from Organization to BillingDetails model

This commit moves everything that's related to billing from
Organization to BillingDetails. This also requires adding
a billing-details adapter, since Ember for some reason fetches
camel-cased endpoints by default.

blakepettersson commented:
Now I've rebased this branch again.

blakepettersson committed:
Merge branch 'master' of github.com:aptible/ember-cli-aptible-shared into extract-billing

Conflicts:
	package.json
